File: ChapServer.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Collections.Concurrent;

namespace ZIM.Server
{
    public class ChapSession
    {
        public string SessionId { get; set; } = "";
        public string ClientId { get; set; } = "";
        public DateTime LastActivity { get; set; }
    }

    public class ChapRequest
    {
        public string? SessionId { get; set; }
        public string? Action { get; set; }
        public JsonElement? Data { get; set; }
    }

    public class ChapResponse
    {
        public bool Success { get; set; }
        public string Message { get; set; } = "";
        public string? NewId { get; set; }
        public object? Data { get; set; }
    }

    public class ChapHandler
    {
        private readonly byte[] _masterKey;
        private readonly ConcurrentDictionary<string, ChapSession> _sessions = new();
        private readonly RandomNumberGenerator _rng = RandomNumberGenerator.Create();

        public ChapHandler(string adminPassword)
        {
            using var sha256 = SHA256.Create();
            _masterKey = sha256.ComputeHash(Encoding.UTF8.GetBytes(adminPassword));
        }

        private byte[] Encrypt(byte[] key, string plaintext)
        {
            using var aes = Aes.Create();
            aes.Key = key;
            aes.Mode = CipherMode.CBC;
            aes.Padding = PaddingMode.PKCS7;
            aes.GenerateIV();

            using var encryptor = aes.CreateEncryptor();
            var plainBytes = Encoding.UTF8.GetBytes(plaintext);
            var cipherBytes = encryptor.TransformFinalBlock(plainBytes, 0, plainBytes.Length);

            var result = new byte[aes.IV.Length + cipherBytes.Length];
            Buffer.BlockCopy(aes.IV, 0, result, 0, aes.IV.Length);
            Buffer.BlockCopy(cipherBytes, 0, result, aes.IV.Length, cipherBytes.Length);
            return result;
        }

        private string? Decrypt(byte[] key, byte[] ciphertextWithIv)
        {
            try
            {
                using var aes = Aes.Create();
                aes.Key = key;
                aes.Mode = CipherMode.CBC;
                aes.Padding = PaddingMode.PKCS7;

                var iv = new byte[16];
                Buffer.BlockCopy(ciphertextWithIv, 0, iv, 0, 16);
                aes.IV = iv;

                var ciphertext = new byte[ciphertextWithIv.Length - 16];
                Buffer.BlockCopy(ciphertextWithIv, 16, ciphertext, 0, ciphertext.Length);

                using var decryptor = aes.CreateDecryptor();
                var plainBytes = decryptor.TransformFinalBlock(ciphertext, 0, ciphertext.Length);
                return Encoding.UTF8.GetString(plainBytes);
            }
            catch
            {
                return null;
            }
        }

        private string GenerateSessionId()
        {
            var bytes = new byte[32];
            _rng.GetBytes(bytes);
            return Convert.ToHexString(bytes);
        }

        private ChapResponse ErrorResponse(string message, string? newId = null)
        {
            return new ChapResponse { Success = false, Message = message, NewId = newId };
        }

        private ChapResponse SuccessResponse(string message, string? newId = null, object? data = null)
        {
            return new ChapResponse { Success = true, Message = message, NewId = newId, Data = data };
        }

        public byte[] HandleLogin(byte[] encryptedData, string clientId)
        {
            var decrypted = Decrypt(_masterKey, encryptedData);
            if (decrypted == null)
            {
                var response = ErrorResponse("authentication_failed");
                return Encrypt(_masterKey, JsonSerializer.Serialize(response));
            }

            var username = decrypted.Trim();
            if (username != "admin")
            {
                var response = ErrorResponse("authentication_failed");
                return Encrypt(_masterKey, JsonSerializer.Serialize(response));
            }

            var sessionId = GenerateSessionId();
            _sessions[sessionId] = new ChapSession
            {
                SessionId = sessionId,
                ClientId = clientId,
                LastActivity = DateTime.UtcNow
            };

            var successResponse = SuccessResponse("login_success", sessionId);
            return Encrypt(_masterKey, JsonSerializer.Serialize(successResponse));
        }

        public byte[] HandleOperation(byte[] encryptedData, string clientId)
        {
            var decrypted = Decrypt(_masterKey, encryptedData);
            if (decrypted == null)
            {
                var response = ErrorResponse("decryption_failed");
                return Encrypt(_masterKey, JsonSerializer.Serialize(response));
            }

            ChapRequest? request;
            try
            {
                request = JsonSerializer.Deserialize<ChapRequest>(decrypted);
            }
            catch
            {
                var response = ErrorResponse("invalid_request");
                return Encrypt(_masterKey, JsonSerializer.Serialize(response));
            }

            if (string.IsNullOrEmpty(request?.SessionId))
            {
                var response = ErrorResponse("missing_session_id");
                return Encrypt(_masterKey, JsonSerializer.Serialize(response));
            }

            if (!_sessions.TryGetValue(request.SessionId, out var session))
            {
                var currentValidId = _sessions.Keys.FirstOrDefault();
                var response = ErrorResponse("session_expired", currentValidId);
                return Encrypt(_masterKey, JsonSerializer.Serialize(response));
            }

            if (session.ClientId != clientId)
            {
                var response = ErrorResponse("client_mismatch");
                return Encrypt(_masterKey, JsonSerializer.Serialize(response));
            }

            _sessions.TryRemove(request.SessionId, out _);

            var newSessionId = GenerateSessionId();
            _sessions[newSessionId] = new ChapSession
            {
                SessionId = newSessionId,
                ClientId = clientId,
                LastActivity = DateTime.UtcNow
            };

            object? operationResult = null;
            if (request.Action != null)
            {
                operationResult = new { action = request.Action, processed = true, timestamp = DateTime.UtcNow };
            }

            var response = SuccessResponse("operation_success", newSessionId, operationResult);
            return Encrypt(_masterKey, JsonSerializer.Serialize(response));
        }

        public bool ValidateSession(string sessionId, string clientId)
        {
            if (!_sessions.TryGetValue(sessionId, out var session))
                return false;
            
            if (session.ClientId != clientId)
                return false;
            
            if (DateTime.UtcNow - session.LastActivity > TimeSpan.FromHours(1))
            {
                _sessions.TryRemove(sessionId, out _);
                return false;
            }
            
            session.LastActivity = DateTime.UtcNow;
            return true;
        }

        public void CleanupExpiredSessions()
        {
            var expired = _sessions
                .Where(s => DateTime.UtcNow - s.Value.LastActivity > TimeSpan.FromHours(1))
                .Select(s => s.Key)
                .ToList();
            
            foreach (var id in expired)
                _sessions.TryRemove(id, out _);
        }
    }
}

File: ChapServer.py
import hashlib
import secrets
import json
from typing import Optional, Dict, Tuple
from datetime import datetime, timedelta
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives.padding import PKCS7
from cryptography.hazmat.backends import default_backend

class ChapSession:
    def __init__(self, session_id: str, client_id: str):
        self.session_id = session_id
        self.client_id = client_id
        self.last_activity = datetime.utcnow()

class ChapHandler:
    def __init__(self, admin_password: str):
        self._master_key = hashlib.sha256(admin_password.encode()).digest()
        self._sessions: Dict[str, ChapSession] = {}
    
    def _encrypt(self, key: bytes, plaintext: str) -> bytes:
        iv = secrets.token_bytes(16)
        cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
        encryptor = cipher.encryptor()
        
        padder = PKCS7(128).padder()
        padded_data = padder.update(plaintext.encode()) + padder.finalize()
        ciphertext = encryptor.update(padded_data) + encryptor.finalize()
        
        return iv + ciphertext
    
    def _decrypt(self, key: bytes, ciphertext_with_iv: bytes) -> Optional[str]:
        try:
            iv = ciphertext_with_iv[:16]
            ciphertext = ciphertext_with_iv[16:]
            
            cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
            decryptor = cipher.decryptor()
            padded_data = decryptor.update(ciphertext) + decryptor.finalize()
            
            unpadder = PKCS7(128).unpadder()
            data = unpadder.update(padded_data) + unpadder.finalize()
            
            return data.decode()
        except Exception:
            return None
    
    def _generate_session_id(self) -> str:
        return secrets.token_hex(32)
    
    def _error_response(self, message: str, new_id: Optional[str] = None) -> dict:
        return {"success": False, "message": message, "newId": new_id}
    
    def _success_response(self, message: str, new_id: Optional[str] = None, data: Optional[dict] = None) -> dict:
        return {"success": True, "message": message, "newId": new_id, "data": data}
    
    def handle_login(self, encrypted_data: bytes, client_id: str) -> bytes:
        decrypted = self._decrypt(self._master_key, encrypted_data)
        if decrypted is None:
            response = self._error_response("authentication_failed")
            return self._encrypt(self._master_key, json.dumps(response))
        
        username = decrypted.strip()
        if username != "admin":
            response = self._error_response("authentication_failed")
            return self._encrypt(self._master_key, json.dumps(response))
        
        session_id = self._generate_session_id()
        self._sessions[session_id] = ChapSession(session_id, client_id)
        
        response = self._success_response("login_success", session_id)
        return self._encrypt(self._master_key, json.dumps(response))
    
    def handle_operation(self, encrypted_data: bytes, client_id: str) -> bytes:
        decrypted = self._decrypt(self._master_key, encrypted_data)
        if decrypted is None:
            response = self._error_response("decryption_failed")
            return self._encrypt(self._master_key, json.dumps(response))
        
        try:
            request = json.loads(decrypted)
        except json.JSONDecodeError:
            response = self._error_response("invalid_request")
            return self._encrypt(self._master_key, json.dumps(response))
        
        session_id = request.get("sessionId")
        if not session_id:
            response = self._error_response("missing_session_id")
            return self._encrypt(self._master_key, json.dumps(response))
        
        session = self._sessions.get(session_id)
        if not session:
            current_valid_id = next(iter(self._sessions.keys())) if self._sessions else None
            response = self._error_response("session_expired", current_valid_id)
            return self._encrypt(self._master_key, json.dumps(response))
        
        if session.client_id != client_id:
            response = self._error_response("client_mismatch")
            return self._encrypt(self._master_key, json.dumps(response))
        
        del self._sessions[session_id]
        
        new_session_id = self._generate_session_id()
        self._sessions[new_session_id] = ChapSession(new_session_id, client_id)
        
        operation_result = None
        action = request.get("action")
        if action:
            operation_result = {
                "action": action,
                "processed": True,
                "timestamp": datetime.utcnow().isoformat()
            }
        
        response = self._success_response("operation_success", new_session_id, operation_result)
        return self._encrypt(self._master_key, json.dumps(response))
    
    def validate_session(self, session_id: str, client_id: str) -> bool:
        session = self._sessions.get(session_id)
        if not session:
            return False
        
        if session.client_id != client_id:
            return False
        
        if datetime.utcnow() - session.last_activity > timedelta(hours=1):
            del self._sessions[session_id]
            return False
        
        session.last_activity = datetime.utcnow()
        return True
    
    def cleanup_expired_sessions(self) -> None:
        now = datetime.utcnow()
        expired = [
            sid for sid, session in self._sessions.items()
            if now - session.last_activity > timedelta(hours=1)
        ]
        for sid in expired:
            del self._sessions[sid]
